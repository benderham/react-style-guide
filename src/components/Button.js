import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';

const StyledButton = styled.button`
  border: solid 1px red;
`;

const Button = ({ label, onClick }) => (
  <StyledButton onClick={onClick}>{label}</StyledButton>
);

Button.propTypes = {
  /** Text shown on the button. */
  label: PropTypes.string.isRequired,
  /** Function passed to the button's 'onClick' event. */
  onClick: PropTypes.func.isRequired,
};

export default Button;
