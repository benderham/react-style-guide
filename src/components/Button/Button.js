import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';

const StyledButton = styled.button`
  border: solid 1px red;
  font-family: ${props => props.theme.fonts.primary};
  font-weight: 600;
  font-size: 1.2rem;
  cursor: pointer;
`;

const Button = ({ disabled, label, onClick, type }) => (
  <StyledButton type={type} onClick={onClick} disabled={disabled}>
    {label}
  </StyledButton>
);

Button.defaultProps = {
  disabled: false,
  type: 'button',
};

Button.propTypes = {
  disabled: PropTypes.bool,
  /** Text shown on the button. */
  label: PropTypes.string.isRequired,
  /** Function passed to the button's 'onClick' event. */
  onClick: PropTypes.func.isRequired,
  type: PropTypes.oneOf(['button', 'reset', 'submit']),
};

export default Button;
