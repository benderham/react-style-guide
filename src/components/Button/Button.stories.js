import React from 'react';
import { storiesOf } from '@storybook/react';
import { action } from '@storybook/addon-actions';
import { withInfo } from '@storybook/addon-info';

import Button from './Button';

storiesOf('Button', module)
  .add(
    'Primary',
    withInfo(`
      Primary button style with text label.
    `)(() => <Button label="Button" onClick={action('button-click')} />),
  )
  .add(
    'Secondary',
    withInfo(`
      Secondary button style with text label.
    `)(() => (
      <Button
        label="Button"
        onClick={action('button-click')}
        kind="secondary"
      />
    )),
  )
  .add(
    'Disabled',
    withInfo(`
      Disabled button.
    `)(() => (
      <Button label="Button" onClick={action('button-click')} disabled />
    )),
  );
