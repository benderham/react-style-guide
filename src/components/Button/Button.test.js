import React from 'react';
import { shallow } from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';
import Button from './Button';

it('should render correctly without error', () => {
  const props = {
    label: 'Button',
    onClick: jest.fn(),
  };
  const output = shallow(<Button {...props} />);
  expect(shallowToJson(output)).toMatchSnapshot();
});

it('should console log error if not passed a label via props', () => {
  const props = {
    onClick: jest.fn(),
  };
  expect(() => <Button {...props} />).toThrow();
});

it('should console log error if not passed an onClick function via props', () => {
  const props = {
    label: 'Button',
  };
  expect(() => <Button {...props} />).toThrow();
});

it(`should have a default type of "button"`, () => {
  const props = {
    label: 'Button',
    onClick: jest.fn(),
  };
  const wrapper = shallow(<Button {...props} />);
  expect(wrapper.props().type).toBe('button');
});

it(`should console log error if passed a type other than "submit", "reset" or "button", via props`, () => {
  const props = {
    label: 'Button',
    onClick: jest.fn(),
    type: 'notsubmit',
  };
  expect(() => <Button {...props} />).toThrow();
});

it(`should have a type of "submit" if passed via props`, () => {
  const props = {
    label: 'Button',
    onClick: jest.fn(),
    type: 'submit',
  };
  const wrapper = shallow(<Button {...props} />);
  expect(wrapper.props().type).toBe('submit');
});

it(`should have a type of "reset" if passed via props`, () => {
  const props = {
    label: 'Button',
    onClick: jest.fn(),
    type: 'reset',
  };
  const wrapper = shallow(<Button {...props} />);
  expect(wrapper.props().type).toBe('reset');
});

it('should handle click event passed via props', () => {
  const mockFn = jest.fn();
  const props = {
    label: 'Button',
    onClick: mockFn,
  };
  const output = shallow(<Button {...props} />);
  output.simulate('click');
  expect(mockFn).toHaveBeenCalled();
});

it('should not be disabled by default', () => {
  const props = {
    label: 'Button',
    onClick: jest.fn(),
  };
  const wrapper = shallow(<Button {...props} />);
  expect(wrapper.props().disabled).toBe(false);
});

it('should be disabled if passed via props', () => {
  const props = {
    label: 'Button',
    onClick: jest.fn(),
    disabled: true,
  };
  const wrapper = shallow(<Button {...props} />);
  expect(wrapper.props().disabled).toBe(true);
});
