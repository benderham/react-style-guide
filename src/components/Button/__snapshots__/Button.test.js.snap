// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render correctly without error 1`] = `
<styled.button
  disabled={false}
  onClick={[Function]}
  type="button"
>
  Button
</styled.button>
`;
